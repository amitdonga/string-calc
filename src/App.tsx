import React, { useState } from "react";
import "./App.css";

function App() {
  const [string, setString] = useState("");
  const [result, setResult] = useState<null | number>(null);
  const [error, setError] = useState("");

  const handleClick = (e: any) => {
    if (string.length === 0) {
      setResult(0);
      return 0;
    }
    let delimiter = /[\n,]/;
    let numbersString = string;

    if (string.startsWith("//")) {
      const delimiterLineEnd = numbersString.indexOf("\\n");
      const customDelimiter = string.slice(2, delimiterLineEnd);
      delimiter = new RegExp(customDelimiter, "g");
      numbersString = string
        .slice(delimiterLineEnd + 2)
        .replace(/\\n/g, customDelimiter);
    } else {
      numbersString = string.replace(/\\n/g, ",");
    }

    const splitNumbers = numbersString
      .split(delimiter)
      .map((num) => num.trim());

    const negatives: number[] = [];
    const sum = splitNumbers.reduce((acc, num) => {
      const value = parseInt(num, 10);
      if (isNaN(value)) {
        return acc;
      }
      if (value < 0) {
        negatives.push(value);
        return acc;
      }
      return acc + value;
    }, 0);

    if (negatives.length > 0) {
      setError(`negative numbers not allowed ${negatives.join(", ")}`);
    }

    setResult(sum);
  };

  return (
    <div className="App">
      <h1>String Calculator</h1>
      <input
        type="text"
        id="string"
        name="string"
        placeholder="Please enter string to calculate"
        onChange={(e) => setString(e.target.value)}
      />
      <button onClick={handleClick}>Calculate</button>
      {result !== null && (
        <div>
          <h3>{`Output: ${result}`}</h3>
        </div>
      )}
      {error && <p style={{ color: "red" }}>{error}</p>}
    </div>
  );
}

export default App;
