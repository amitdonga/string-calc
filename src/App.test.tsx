import React, { act } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";

describe("String Calculator", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const user = userEvent;

  test("should renders component", () => {
    render(<App />);
    const linkElement = screen.getByRole("heading", {
      level: 1,
      name: "String Calculator",
    });
    expect(linkElement).toBeInTheDocument();
  });

  test("should renders component and match snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test("should renders component and all loaded element", () => {
    render(<App />);
    const btnElement = screen.getByRole("button", { name: "Calculate" });
    const inputElement = screen.getByPlaceholderText(
      "Please enter string to calculate"
    );
    expect(inputElement).toBeInTheDocument();
    expect(btnElement).toBeInTheDocument();
  });

  test("should renders component and click calculate without enter value", async () => {
    render(<App />);
    await act(async () => {
      const btnElement = screen.getByRole("button", { name: "Calculate" });
      await user.click(btnElement);
    });
    const outputElement = screen.getByText("Output: 0");
    expect(outputElement).toBeInTheDocument();
  });

  test("should renders component and enter string and validate the output", async () => {
    render(<App />);
    await act(async () => {
      const btnElement = screen.getByRole("button", { name: "Calculate" });
      const inputElement = screen.getByPlaceholderText(
        "Please enter string to calculate"
      );
      await user.type(inputElement, "1,2,3");
      await user.click(btnElement);
    });

    const outputElement = screen.getByText("Output: 6");
    expect(outputElement).toBeInTheDocument();
  });

  test("should renders component and enter string with \n and other delimiter and validate the output", async () => {
    render(<App />);
    await act(async () => {
      const btnElement = screen.getByRole("button", { name: "Calculate" });
      const inputElement = screen.getByPlaceholderText(
        "Please enter string to calculate"
      );
      await user.type(inputElement, "//;\\n1;2\\n3;4\\n4;5");
      await user.click(btnElement);
    });

    await new Promise((resolve) => {
      setTimeout(resolve, 100);
    });
    const outputElement = await screen.findByText("Output: 19");
    expect(outputElement).toBeInTheDocument();
  });

  test("should renders component and enter string with negative number and validate the output", async () => {
    render(<App />);
    await act(async () => {
      const btnElement = screen.getByRole("button", { name: "Calculate" });
      const inputElement = screen.getByPlaceholderText(
        "Please enter string to calculate"
      );
      await user.type(inputElement, "//;\\n1;2\\n3;4\\n4;5;-4;-3");
      await user.click(btnElement);
    });

    await new Promise((resolve) => {
      setTimeout(resolve, 100);
    });
    const outputElement = await screen.findByText("Output: 19");
    expect(outputElement).toBeInTheDocument();
    const errorElement = await screen.findByText(
      "negative numbers not allowed -4, -3"
    );
    expect(errorElement).toBeInTheDocument();
  });
});
