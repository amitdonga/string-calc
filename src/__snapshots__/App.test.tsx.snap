// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String Calculator should renders component and match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      String Calculator
    </h1>
    <input
      id="string"
      name="string"
      placeholder="Please enter string to calculate"
      type="text"
    />
    <button>
      Calculate
    </button>
  </div>
</div>
`;
